src/mammos_parser/util.py

```
@@ -0,0 +1,211 @@
+    files/directories found during file collection set required_elements_found=False.
+    """
+
+    root_dir: Path
+    tree_ok: bool
+    collected_files: set[str]
+    collected_dirs: set[str]
+
+    def __bool__(self):
+        """Instance is true if all required files/dirs were found."""
+        return self.tree_ok
+
+    def __add__(self, other: Self):
+        """Combine two Collected instances."""
+        if self.root_dir != other.root_dir:
+            raise ValueError("Different root directories.")
+        return self.__class__(
+            self.root_dir,
+            self.tree_ok and other.tree_ok,
+            self.collected_files | other.collected_files,
+            self.collected_dirs | other.collected_dirs,
+        )
+
+
+def check_directory(
+    root_dir: Path,
+    dir_name: str | Path,
+    *,
+    required_files: set[str] | None = None,
+    optional_files: set[str] | None = None,
+    required_files_from_choices: list[set[str]] | None = None,
+    required_file_pairs: list[tuple[str, str]] | None = None,
+    required_subdirs: set[str] | None = None,
+    optional_subdirs: set[str] | None = None,
+) -> Collected:
+    """Check content of directory.
+
+    Args:
+        root_dir: Path to dataset root, must be absolute.
+        dir_name: Directory to be parsed, relative to `root_dir`.
+        required_files: List of required files in `dir_name`.
+        optional_files: List of optional files in `dir_name`.
+        required_files_from_choices: Each element in the list defines a set of files
+            out of which exactly one must exist.
+        required_file_pairs: Each element in the list defines two prefixes for which
+            files with a common suffix must exist. Example: for the element
+            ``("a-", "b-")`` paires like ``a-0`` AND ``b-0`` or ``a-in.txt`` AND
+            ``b-in.txt`` must exist.
+        required_subdirs: List of subdirectories (only one level) that must exist in
```

Why only one level? Required files can be checked across multiple levels.

---

I do not trust the current implementation to work across multiple levels of directories (and we don't really need it for the current use case, therefor I did not spend time on it). The implementation does not enforce checking only one level of subdirectories but I would highly recommend it. We could even enforce that for now.

---

As we do not currently need this - I am okay to not to implement checking multiple levels. I would have a slight preference for us to check and raise an error if someone tries rather than it maybe/working/silently failing. That way it also reminds us to actually update and test this functionality if we need it in the future.

---

Correction: the current implementation does only read subdirectories of the checked directories but does not recurse into these. Requesting nested subdirectories will therefor fail because we do checks on a string level. Example: we have a directory `mysub/nested`, and the user asks for `"mysub/nested"`. The code only picks up `"mysub"` and the comparison will fail, reporting `"mysub/nested"` as missing and `"mysub"` as unexpected.